azext_iot/dps/providers/discovery.py

```diff
@@ -0,0 +1,68 @@
+# coding=utf-8
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from knack.log import get_logger
+from azure.cli.core.commands.client_factory import get_subscription_id
+from azext_iot.common._azure import CONN_STR_TEMPLATE
+from azext_iot.common.base_discovery import BaseDiscovery
+from azext_iot.common.shared import DiscoveryResourceType
+from azext_iot._factory import iot_service_provisioning_factory
+from typing import Dict
+
+logger = get_logger(__name__)
+
+
+class DPSDiscovery(BaseDiscovery):
+    def __init__(self, cmd):
+        super().__init__(
+            cmd,
+            DiscoveryResourceType.DPS.value
+        )
+
+    def _initialize_client(self):
+        if not self.client:
+            if getattr(self.cmd, "cli_ctx", None):
+                self.client = iot_service_provisioning_factory(self.cmd.cli_ctx).iot_dps_resource
+                self.sub_id = get_subscription_id(self.cmd.cli_ctx)
+            else:
+                self.client = self.cmd
+
+            # Weirdness thanks to client discrepencies
```

Comment:
Valuable to explain more in this comment so when a dev that doesn't have context looks at it they have a clearer understanding.